langkit/ocaml_api.py

```diff
@@ -0,0 +1,444 @@
+from __future__ import absolute_import, division, print_function
+
+import langkit.compiled_types as ct
+from langkit.compiled_types import T
+from langkit.language_api import AbstractAPISettings
+from langkit.utils import dispatch_on_type
+
+
+class OCamlAPISettings(AbstractAPISettings):
+    """Container for OCaml API generation settings."""
+
+    name = 'ocaml'
+
+    def __init__(self, ctx, c_api_settings):
+        self.context = ctx
+        self.c_api_settings = c_api_settings
+
+    def is_empty_type(self, type):
+        """
+        Test if the given type would end-up being empty, so that we can avoid
+        generating empty types.
+
+        :param ct.CompiledType: The type we want to check if it is empty.
+        :rtype: bool
+        """
+        return dispatch_on_type(type, [
+            (T.Bool, lambda _: False),
+            (T.Int, lambda _: False),
+            (T.Character, lambda _: False),
+            (T.Token, lambda t: False),
+            (T.Symbol, lambda _: False),
+            (ct.EnumType, lambda t: False),
+            (ct.ASTNodeType, lambda t:
+                len(t.concrete_subclasses) == 0),
+            (ct.EntityType, lambda t: self.is_empty_type(t.astnode)),
+            (T.AnalysisUnit, lambda _: False),
+            (ct.ArrayType, lambda t: self.is_empty_type(t.element_type)),
+            (ct.StructType, lambda t:
+                len([
+                    t
```

Comment:
It looks like you want to check if there is no field that has an empty type ? I'd recommend in that case writing:

```python
not any(self.is_empty_type(f.type) for f in t.get_fields())
```

Also nit: strange formatting, I'd write

```python
                    t for field in t.get_fields()
                    if not self.is_empty_type(field.type)
```

Reply:
I concur. By the way, could you enhance the docstring to describe what an empty type is exactly? Looking at the implementation, it looks like it will return False for every type.

Reply:
I will update the comment. By "empty type" I mean empty struct, or struct that only contains empty structs.
In the implementation, the function returns true when a struct type does not have any field, or the only fields that it has are also empty structs.

Reply:
Interesting… Does this take into account inherited fields? (for nodes, which are also structs) If nodes with only inherited fields are not empty, then we probably never had any empty struct. I guess it’s allowed, but since it’s never been tested I would not bet it works. ;-)

Reply:
nodes are handled separately and are considered empty if they do not have any concrete subclasses. Even if a node has no field, we still have the c_value so the OCaml struct is not empty. (I might have misunderstood your concern)

Reply:
No, you’re fine. :-)

The case you handle here (structs with no field and abstract nodes with no concrete subclass) seem quite pathological, and I’m tempted to make sure they never happen in practice (i.e. make them illegal for Langkit). That would enable you to simplify your change. What do you think?

Reply:
Yes, I had to make this because it happened in some of langkit's tests.
One example is the Metadata structure that sometimes does not contain any field.
I also saw some ASTNode without any subclasses in the testsuite.

It would definitely simplify some code since I would not need to check whether an empty OCaml type would be generated.

Reply:
Oh right, I forgot about metadata: I’m taking this back (we’ll keep supporting empty structs). And I just remembered that we need to support abstract classes with no concrete subclass because of the way our generic list type works. So never mind :-) Thank you for updating the docstring, then!

Note that you could simplify the call to dispatch_on_type: for instance replacing all cases that unconditionally return False with a single `(ct.CompiledType, lambda _: False)` clause at the end of the list.